src/PushPackage.php

```diff
@@ -0,0 +1,236 @@
+     */
+    protected $package;
+
+    /**
+     * The destination folder in which to unzip the package.
+     *
+     * @var string
+     */
+    protected $destinationFolder;
+
+    /**
+     * The authentication for the ssh connection.
+     *
+     * @var AbstractAuth
+     */
+    protected $auth;
+
+    /**
+     * The fully qualified classname of the scp factory.
+     *
+     * @var string
+     */
+    protected $scpFactory = ScpPhpseclibFactory::class;
+
+
+    /**
+     * The fully qualified classname of the ssh factory.
+     *
+     * @var string
+     */
+    protected $sshFactory = SshPhpseclibFactory::class;
+
+    /**
+     * Creates a new Scp task.
+     *
+     * @param string $host
+     *   The host.
+     * @param AbstractAuth $auth
+     *   Authentication data.
+     */
+    public function __construct($host, AbstractAuth $auth)
+    {
+        $this->host = $host;
+        $this->auth = $auth;
+    }
+
+    /**
+     * Sets the port to connect on.
+     *
+     * @param int $port
+     *   The port to connect on.
+     *
+     * @return $this
+     */
+    public function port($port)
+    {
+        $this->port = $port;
+
+        return $this;
+    }
+
+    /**
+     * Sets the ssh timeout.
+     *
+     * @param int $timeout
+     *   Timeout in seconds.
+     *
+     * @return $this
+     */
+    public function timeout($timeout)
+    {
+        $this->timeout = $timeout;
+
+        return $this;
+    }
+
+    /**
+     * Sets the package path.
+     *
+     * @param string $path
+     *   The package path.
+     *
+     * @return $this
+     */
+    public function package($path)
+    {
+        $this->package = $path;
+
+        return $this;
+    }
+
+    /**
+     * Sets the destination folder in which to unzip the package on the server.
+     *
+     * @param string $destinationFolder
+     *   The destination folder.
+     *
+     * @return $this
+     */
+    public function destinationFolder($destinationFolder)
+    {
+        $this->destinationFolder = $destinationFolder;
+
+        return $this;
+    }
+
+    /**
+     * Set the ScpFactory.
+     *
+     * @param sting|ScpFactoryInterface $class
+     *   A factory instance or the fully qualified classname of the scp factory.
+     *   The given class (whether it's a classname or instance) must implement
+     *   \DigipolisGent\Robo\Task\Deploy\Scp\Factory\ScpFactoryInterface.
+     *
+     * @throws InvalidArgumentException
+     *   If the class is not an instance of
+     *   \DigipolisGent\Robo\Task\Deploy\Scp\Factory\ScpFactoryInterface.
+     *
+     * @return $this
+     */
+    public function scpFactory($class)
+    {
+        if (!is_subclass_of($class, ScpFactoryInterface::class)) {
+            throw new InvalidArgumentException(sprintf(
+                'SCP Factory %s does not implement %s.',
+                $class,
+                ScpFactoryInterface::class
+            ));
+        }
+        $this->scpFactory = $class;
```

Comment:
All code using the factory should work with both a string and object: is_subclass_of works and call_user_func_array works with either strings or objects.